pkg/storage/scrub.go

```diff
@@ -0,0 +1,244 @@
+	"github.com/opencontainers/umoci/oci/casext"
+)
+
+const (
+	imageNameWidth = 32
+	tagWidth       = 24
+	statusWidth    = 8
+	errorWidth     = 8
+
+	colImageNameIndex = 0
+	colTagIndex       = 1
+	colStatusIndex    = 2
+	colErrorIndex     = 3
+)
+
+type ScrubImageResult struct {
+	ImageName string `json:"image_name"`
+	Tag       string `json:"tag"`
+	Status    string `json:"status"`
+	Error     string `json:"error"`
+}
+
+type ScrubResults struct {
+	ScrubResults []ScrubImageResult `json:"scrub_results"`
+}
+
+func (sc StoreController) CheckAllBlobsIntegrity() (ScrubResults, error) {
+	imagesList := []ImageStore{}
+	imagesList = append(imagesList, sc.DefaultStore)
+
+	for _, subStore := range sc.SubStore {
+		imagesList = append(imagesList, subStore)
+	}
+
+	results := ScrubResults{}
+
+	for _, is := range imagesList {
+		images, err := is.GetRepositories()
+
+		if err != nil {
+			return results, err
+		}
+
+		for _, repo := range images {
+			imageResults, err := checkImage(repo, is)
+
+			if err != nil {
+				return results, err
+			}
+
+			results.ScrubResults = append(results.ScrubResults, imageResults...)
+		}
+	}
+
+	return results, nil
+}
+
+func checkImage(imageName string, is ImageStore) ([]ScrubImageResult, error) {
+	results := []ScrubImageResult{}
+
+	dir := path.Join(is.RootDir(), imageName)
+	if !is.DirExists(dir) {
+		return results, errors.ErrRepoNotFound
+	}
+
+	ctxUmoci := context.Background()
+
+	oci, err := umoci.OpenLayout(dir)
+	if err != nil {
+		return results, err
+	}
+
+	defer oci.Close()
+
+	is.RLock()
+	defer is.RUnlock()
+
+	buf, err := ioutil.ReadFile(path.Join(dir, "index.json"))
+
+	if err != nil {
+		//is.log.Error().Err(err).Str("dir", dir).Msg("failed to read index.json")
+		return results, err
+	}
+
+	var index ispec.Index
+	if err := json.Unmarshal(buf, &index); err != nil {
+		//is.log.Error().Err(err).Str("dir", dir).Msg("invalid JSON")
+		return results, errors.ErrRepoNotFound
+	}
+
+	for _, m := range index.Manifests {
+		tag, ok := m.Annotations[ispec.AnnotationRefName]
+		if ok {
+			imageResult := checkIntegrity(ctxUmoci, imageName, tag, oci, m, dir)
+			results = append(results, imageResult)
+		}
+	}
+
+	return results, nil
+}
+
+func checkIntegrity(ctx context.Context, imageName, tagName string, oci casext.Engine, manifest ispec.Descriptor,
+	dir string) ScrubImageResult {
+	// check manifest and config
+	stat, err := umoci.Stat(ctx, oci, manifest)
+
+	imageRes := ScrubImageResult{}
+
+	if err != nil {
+		imageRes = getResult(imageName, tagName, err)
+	} else {
+		// check layers
+		for _, s := range stat.History {
+			layer := s.Layer
+			if layer == nil {
+				continue
+			}
+
+			// check if layer exists
+			bytes, err := ioutil.ReadFile(path.Join(dir, "blobs/sha256", layer.Digest.Hex()))
+			if err != nil {
+				imageRes = getResult(imageName, tagName, errors.ErrBlobNotFound)
+				break
+			}
+
+			// check layer content
+			expectedHash := fmt.Sprintf("%x", sha256.Sum256(bytes))
```

Comment:
Should not refer to any specific hash algorithm (sha256) here in this code directly.

Reply:
Because I'm using the above pattern suggested by you for checking layers, I don't need anymore this part of code so I removed it.